Add Performant Translations

swissspidy opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/59656

**What it does**

* If an MO file has a corresponding PHP file, that file is loaded instead.

**What it does not do** (notable changes from the plugin)

* **Does not** automatically convert any MO files to PHP files upon reading.
	* This avoids any FS interaction on regular requests, avoiding any unexpected results.
	* Makes the logic much simpler.
	* This additional feature can stay in the Performant Translations plugin for those who wish to use it.
* **No** integration with `Language_Pack_Upgrader` to automatically convert MO files to PHP files.
	* This was only needed because WordPress.org didn't yet serve PHP files in language packs.
	* If someone is using a custom translation platform like [Traduttore](https://github.com/wearerequired/traduttore), it is up to them to provide PHP files as well (if they want to)

**Related**

* [PR for WP-CLI's i18n command](https://github.com/wp-cli/i18n-command/pull/363)
* [PR for adding PHP format support in GlotPress](https://github.com/GlotPress/GlotPress/pull/1626)
* [Meta Ticket & Patch adding PHP files to language packs](https://meta.trac.wordpress.org/ticket/7296)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

swissspidy committed:
Add Ginger MO

swissspidy committed:
Merge branch 'WordPress:trunk' into add/ginger-mo

swissspidy commented on src/wp-admin/includes/class-language-pack-upgrader.php:
```diff
@@ -288,6 +288,45 @@ public function bulk_upgrade( $language_updates = array(), $args = array() ) {
 			)
 		);
 
+		foreach ( $language_updates_results as $translation ) {
```
Still needed? If so, check first if language pack already contains PHP files or not.

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Fix conditionals

swissspidy committed:
Add realpath

swissspidy committed:
Add PHPUnit tests

swissspidy committed:
Lint fixes

swissspidy committed:
Check if is string

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Remove json support and upgrader integration

swissspidy committed:
Remove some more files

swissspidy committed:
Remove unlink

swissspidy committed:
Delete PHP translation file on plugin uninstall

swissspidy committed:
Add back one file that’s actually used

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
docblock updates

swissspidy committed:
More doc updates

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Rename classes

swissspidy committed:
Rename var

swissspidy committed:
Fix alignment

swissspidy committed:
Fix test

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Merge latest change from plugin

See swissspidy/performant-translations#115

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
-		$mo->merge_with( $l10n[ $domain ] );
+	$mofile_preferred = str_replace( '.mo', ".mo.$preferred_format", $mofile );
+
+	if ( 'mo' !== $preferred_format || str_ends_with( $mofile, $preferred_format ) ) {
```
What's the purpose of the `str_ends_with()` here? Not sure I understand. From my understanding, it makes sense to enter this clause if the intended format is _not_ `mo` (as it would be today, since it's the only option). But why the second condition? I would expect the `$mofile` variable at this point to end in `mo`, so it would enter this clause if `$preferred_format` was `mo`. Why is that needed, shouldn't it just go with the other logic tree instead (which is more similar to what's happening today?

swissspidy replied:
> But why the second condition?

This is if someone calls `load_textdomain()` with the path to a PHP file. Yes, I've seen it happen. It's some logic from the plugin to improve Loco Translate compatibility. See https://github.com/swissspidy/performant-translations/pull/93

Now that file creation is no longer part of this patch I can remove it though!

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
 
-	if ( isset( $l10n[ $domain ] ) ) {
-		$mo->merge_with( $l10n[ $domain ] );
+	$mofile_preferred = str_replace( '.mo', ".mo.$preferred_format", $mofile );
```
Nit-pick: As far as I can tell this variable is only used in the if clause below, so this definition should probably be moved there.

swissspidy replied:
Ah yes, wasn't originally the case. Only happened to be the case after a refactor. Good catch.

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+	if ( 'mo' !== $preferred_format || str_ends_with( $mofile, $preferred_format ) ) {
+
+		/** This filter is documented in wp-includes/l10n.php */
+		$mofile_preferred = apply_filters( 'load_textdomain_mofile', $mofile_preferred, $domain );
```
Why does this filter need to fire again here? I'm confused by this for two reasons:
* It's already applied in the same function further up.
* It would now potentially receive a `.mo.php` file path, which may break existing use of that filter.

swissspidy replied:
Related to your other question:

> Similar to the other new filter, what's the rationale / use-case for introducing the `load_translation_file` filter?

I was thinking a long time about these two filters and what the best approach is.

`load_textdomain_mofile` is historically the one filter to change the file path, but plugins expect only MO files here.

Hence the introduction of the more generic `load_translation_file` filter which could also receive a PHP file.

I still opted to call the `load_textdomain_mofile` filter for PHP files as well, but that's probably not the safest approach. So I'll just remove it again.

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+	WP_I18n_Translation_Controller::instance()->set_locale( $locale );
 
-		return false;
+	/**
+	 * Filters the preferred file format for translation files.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $preferred_format Preferred file format. Possible values: 'php', 'mo'. Default: 'php'.
+	 */
+	$preferred_format = apply_filters( 'translation_file_format', 'php' );
+	if ( ! in_array( $preferred_format, array( 'php', 'mo' ), true ) ) {
+		$preferred_format = 'php';
 	}
```
What's the rationale for introducing this filter? What would be a realistic use-case for it? If there is one, I'm not opposed, but I would advise to rather be cautious on introducing new filters when we don't have to as it limits what changes we can make later without breaking BC.

swissspidy replied:
This allows people to opt out of the behavior of preferring PHP files over MO files, simple as that.

100% sure there will be people who want to do that, so they need a way to do that.

felixarntz replied:
@swissspidy Why would people want to do that? In what case would MO files have a benefit over PHP? I just want to make sure we allow it for good reasons, not just because someone wants that (without a good reason).

The other related question: How would that work with shipping translations? Wouldn't this preference need to equally apply when eventually getting language packs from .org (i.e. a parameter needs to be sent to .org which format is preferred)?

swissspidy replied:
Dotorg sends both formats in a ZIP file, it does not support conditionally sending only half.

Originally the filter was added to choose between MO/JSON/PHP. Now someone might want to prefer PHP if they have some concerns about that format or if they are not even using dotorg and only have MO files anyway, skipping a lot of file_exists checks.

Don‘t have a string opinion though, we can also remove and add back later if needed 🤷‍♂️

felixarntz replied:
I would prefer that, we can always introduce filters when needed, but we can't remove them. I prefer that over adding filters for a _potential_ need. It helps avoid unnecessary filters that we then still have to maintain forever.

The `file_exists()` checks is a good point, though I'd still prefer to remove for now and park this for a later conversation.

swissspidy replied:
It would be relevant for https://github.com/WordPress/wordpress-develop/pull/5306#issuecomment-1805349307, so let's await that discussion...

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+		 * @param string $file   Path to the translation file to load.
+		 * @param string $domain The text domain.
+		 */
+		$mofile_preferred = apply_filters( 'load_translation_file', $mofile_preferred, $domain );
```
Similar to the other new filter, what's the rationale / use-case for introducing this filter?

swissspidy replied:
Let's combine the threads

felixarntz replied:
@swissspidy It's another use-case though, I don't think you address this on the other filter thread. Or maybe I missed it?

swissspidy replied:
So this is just like the load_textdomain_mofile filter, but the file being passed could also be a PHP file. For back compat I‘m not reusing the same filter.

felixarntz replied:
Makes sense, let's keep this one for parity.

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
 
-	unset( $l10n_unloaded[ $domain ] );
+	/** This action is documented in wp-includes/l10n.php */
+	do_action( 'load_textdomain', $domain, $mofile );
+
+	/** This filter is documented in wp-includes/l10n.php */
+	$mofile = apply_filters( 'load_textdomain_mofile', $mofile, $domain );
+
+	/** This filter is documented in wp-includes/l10n.php */
+	$mofile = apply_filters( 'load_translation_file', $mofile, $domain );
```
A bit related to my other question above, why are all these filters _rerun_ here?

swissspidy replied:
Ah, good catch, a bit of a leftover from when this was added via a filter.

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+	/** This action is documented in wp-includes/l10n.php */
+	do_action( 'load_textdomain', $domain, $mofile );
+
+	/** This filter is documented in wp-includes/l10n.php */
+	$mofile = apply_filters( 'load_textdomain_mofile', $mofile, $domain );
+
+	/** This filter is documented in wp-includes/l10n.php */
+	$mofile = apply_filters( 'load_translation_file', $mofile, $domain );
+
+	$success = WP_I18n_Translation_Controller::instance()->load( $mofile, $domain, $locale );
+
+	if ( $success ) {
+		if ( isset( $l10n[ $domain ] ) && $l10n[ $domain ] instanceof MO ) {
+			WP_I18n_Translation_Controller::instance()->load( $l10n[ $domain ]->get_filename(), $domain, $locale );
+		}
 
-	$l10n[ $domain ] = &$mo;
+		// Unset Noop_Translations reference in get_translations_for_domain.
+		unset( $l10n[ $domain ] );
 
-	$wp_textdomain_registry->set( $domain, $locale, dirname( $mofile ) );
+		$l10n[ $domain ] = new WP_I18n_Translations( $domain );
+
+		$wp_textdomain_registry->set( $domain, $locale, dirname( $mofile ) );
+	}
```
Most of this looks very much like duplicate code of what is in the if clause, with some tiny adjustments like probably the variables. Can we move this to a separate function (`@access private` would be okay) to avoid that?

swissspidy replied:
It's only a handful of lines, not really worth creating another function for this IMHO.

felixarntz replied:
Reviewing the code more closely, I don't even think we need a new function, we just need to simplify the code. I don't see a good reason for all these lines being duplicated. As far as I can tell, they are _exactly_ the same, just acting on a different file. We can probably refactor with a `foreach` loop to avoid the duplicate code. I'll post a suggestion.

felixarntz replied:
I just realized I can't leave a suggestion on that code because GH doesn't allow it, so just sharing here:
* We can use an array of files to attempt and then avoid this duplicate code by using a `foreach` loop.
* We run the logic for every file in the array (either the PHP file and then the MO file, or just the MO file), and as soon as one file was loaded we return `true`. Or, if for some reason nothing is loaded, we return `true` at the end, which would be the same as today.

Something like:
```php
$translation_files = array( $mofile );
if ( 'mo' !== $preferred_format ) {
	array_unshift(
		$translation_files,
		str_replace( '.mo', ".mo.$preferred_format", $mofile )
	);
}

foreach ( $translation_files as $translation_file ) {
	// All logic currently duplicated, using $translation_file instead of $mofile / $mofile_preferred.
	// Including `return true` on success.
}

// Only relevant if no file was actually loaded, in which case today the function also still returns true.
return true;
```

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
-	if ( ! $mo->import_from_file( $mofile ) ) {
-		$wp_textdomain_registry->set( $domain, $locale, false );
+	// Ensures the correct locale is set as the current one, in case it was filtered.
+	WP_I18n_Translation_Controller::instance()->set_locale( $locale );
```
Nit-pick: Since the controller is accessed several times in this function, it would be a bit more readable to assign `WP_I18n_Translation_Controller::instance()` to a variable.

felixarntz commented on src/wp-includes/i18n/class-wp-i18n-translations.php:
```diff
@@ -0,0 +1,158 @@
+		return null;
+	}
+
+	/**
+	 * Magic setter.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $name Property name.
+	 * @param mixed  $value Property value.
+	 */
+	public function __set( string $name, $value ) {}
```
Why is this needed when it doesn't set anything?

felixarntz commented on src/wp-includes/i18n/class-wp-i18n-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string                            $textdomain Text domain.
+	 * @param WP_I18n_Translation_File|string $mo         Translation file instance or file name.
+	 * @param string                            $locale     Optional. Locale. Default all locales.
```
Indentation is off here.

felixarntz commented on src/wp-includes/i18n/class-wp-i18n-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string                            $textdomain Text domain.
```
```suggestion
	 * @param string                            $textdomain Optional. Text domain. Default 'default'.
```

felixarntz commented on src/wp-includes/i18n/class-wp-i18n-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $textdomain Text domain.
```
```suggestion
	 * @param string $textdomain Optional. Text domain. Default 'default'.
```

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -866,8 +925,9 @@ function unload_textdomain( $domain, $reloadable = false ) {
 		unset( $l10n[ $domain ] );
 
+		// Since we support multiple locales, we don't actually need to unload reloadable text domains.
 		if ( ! $reloadable ) {
-			$l10n_unloaded[ $domain ] = true;
+			return WP_I18n_Translation_Controller::instance()->unload( $domain );
 		}
```
I am seeing throughout this PR you no longer set `$l10n_unloaded[ $domain ] = true;`, also via `unload_textdomain( ..., true )`. Does that effectively mean we never unload any text domains? What defines whether a text domain is reloadable or not?

swissspidy replied:
`$l10n_unloaded[ $domain ] = true;` is still called in some places.

> What defines whether a text domain is reloadable or not?

This is up to the caller of `unload_textdomain()`, `$reloadable` is the second 

The whole `$reloadable` concept is related to the locale switching feature in WordPress. The problem there was that WordPress could only load 1 translation at a time. So when switching locales, translations had to be unloaded but you'd still need to be able to load them again just-in-time. There's also test coverage around this logic.

felixarntz replied:
@swissspidy I'm not sure this answers my question. I understand the function signature, but what's a use-case where a locale should _not_ be reloadable?

swissspidy replied:
I don‘t know off the top of my head if we use that in core, but some plugins might want to force unload without it being able to be loaded again accidentally. I‘d need to dig up the old ticket to find examples.

swissspidy replied:
btw I just restored that line in question here for better back compat with that global.

felixarntz replied:
> btw I just restored that line in question here for better back compat with that global.

Is that correct though? For example, in `unload_textdomain()`, the `$l10n_unloaded[ $domain ] = true;` is only applied if the translation is _not_ reloadable. But as far as I can tell, per the logic here it's always reloadable, so I think we would instead need to `unset( $l10n_unloaded[ $domain ] );`, if anything.

felixarntz replied:
> I don‘t know off the top of my head if we use that in core, but some plugins might want to force unload without it being able to be loaded again accidentally. I‘d need to dig up the old ticket to find examples.

That would be good. I mostly want to know _why_. _Why_ would someone want to prevent a translation file from being reloaded?

swissspidy replied:
OK just looked at it again. So, `unload_textdomain()` used to unload textdomains completely, preventing them from being reloaded again. That was and is the default behavior. Then we introduced locale switching.

And when we switch locales, we used to need to unload existing translations, load new ones, and then unload those again and load the original ones again. Hence we added the new `$reloadable` param to support this in a back compat way.

Now while we still need to support the original behavior, `$reloadable` becomes easier because we don't actually unload in that case, we can just store multiple languages in memory.

Makes sense?

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+			// Unset Noop_Translations reference in get_translations_for_domain.
+			unset( $l10n[ $domain ] );
+
+			$l10n[ $domain ] = new WP_I18n_Translations( $domain );
```
So as far as I can tell, this `WP_I18n_Translations` class replaces the `MO` class? If that's the case, two related questions:
1. Is its interface fully compatible (to avoid potential BC breakage)?
2. Is there any remaining value for the `MO` class or could it effectively be deprecated?

swissspidy replied:
Yes it does.

1. Not 100%, but it covers the most common API touchpoints.
	- I haven't found any instances where more changes were required. Be it through tests or wpdirectory.net.
	- Actually extending `Translations` or the `MO` class is not possible because then the magic getter for `entries` and `headers` wouldn't work—unless we make changes to those classes.
	- I'd imagine getting to full compatibility gains only little safety for a lot of work.
2. Yes, there is. No, we will likely never be able to deprecate it. Many projects will continue to have MO files but no PHP files, and we need to support that. Plus, users getting deprecation warnings for no reason is not great. Second reason is that plugins like GlotPress directly use these classes.

swissspidy replied:
See also https://github.com/swissspidy/performant-translations/issues/43

felixarntz replied:
@swissspidy Not sure I agree on the 2nd point.
* Projects with only MO files can still handle them with the new class responsible for MO files.
* Deprecation warnings aren't for no reason. If there's code which is no longer used, it's a maintenance burden to keep it around. We don't delete code because that could introduce BC breakage, but deprecating unused code IMO is the right path as it encourages them to update. Plugins that use the `MO` class could gradually update to use `WP_I18n_Translations` / `WP_I18n_Translation_File_MO` - without any rush, as WordPress deprecations never get removed. But they won't need to be explicitly maintained anymore either.

swissspidy replied:
The MO class from the polo library as used by GlotPress can support many more things that are not covered by our class. I really don‘t think we can deprecate it. But let‘s ask @ocean90

swissspidy replied:
From what I can see, GlotPress mostly directly uses the `Translations` and `Translation_Entry` classes. These two and `MO` are all from the same pomo library, which is theoretically an external library. It would be weird to deprecate only one of those. So as an i18n component maintainer I am -1 on this suggestion.

felixarntz reviewed (commented):
@swissspidy There is a lot to review here and I am not super familiar with the underlying codebase, so my initial review consists mostly of questions (and some little code improvements).

I'm excited to see this materialize, and am hoping to improve my understanding of this API in the process :)

swissspidy committed:
Fix alignment

audrasjb commented on src/wp-includes/i18n/class-wp-i18n-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	}
+
+	/**
+	 * Unload all translation files or a specific one for a given text domain.
```
As per docblock standards, let's use third-person singular verb: 
`Unloads all translation files or a specific one for a given text domain.`

audrasjb commented on src/wp-includes/i18n/class-wp-i18n-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param array{0: string, 1: string} $plurals    Pair of singular and plural translation.
```
"Pair of singular and plural translation**S**" ?

audrasjb commented on src/wp-includes/i18n/class-wp-i18n-translations.php:
```diff
@@ -0,0 +1,158 @@
+	public function __set( string $name, $value ) {}
+
+	/**
+	 * Build a Translation_Entry from original string and translation strings.
```
Builds

audrasjb reviewed (commented):
Can't wait to see this merged in core, thanks!
I left a few very minor docblock nitpicks.

swissspidy committed:
Apply some suggestions from code review

ocean90 commented on src/wp-admin/includes/plugin.php:
```diff
@@ -1009,6 +1009,7 @@ function delete_plugins( $plugins, $deprecated = '' ) {
 			foreach ( $translations as $translation => $data ) {
 				$wp_filesystem->delete( WP_LANG_DIR . '/plugins/' . $plugin_slug . '-' . $translation . '.po' );
 				$wp_filesystem->delete( WP_LANG_DIR . '/plugins/' . $plugin_slug . '-' . $translation . '.mo' );
+				$wp_filesystem->delete( WP_LANG_DIR . '/plugins/' . $plugin_slug . '-' . $translation . '.mo.php' );
```
What's the reason for using .mo.php instead of just .php?

swissspidy replied:
Glad you asked!

Turns out WP always tries to load `{$locale}.php` if it exists:

https://github.com/WordPress/wordpress-develop/blob/d088e31c73456179502c9bd5354fc43c6267bd7a/src/wp-settings.php#L519-L524

So this different file extension is needed to not break older sites.

See https://github.com/swissspidy/performant-translations/issues/77

ocean90 replied:
Ah! It's just odd to always keep the .mo in the name in case we want to add new formats.

Just thinking out load: What other formats in PHP we could use for the translations? Maybe a serialized string? Highly unlikely but still. An object? What about naming the current format `array.php` so we would have `de_DE.array.php` and maybe `de_DE.serialized.php` or `de_DE.object.php`?

swissspidy replied:
The thought behind `.mo.php` was that it makes it immediately obvious that it's related to translations, like `.l10n.php`. But if you want to split by format we can do `.array.php` too.

swissspidy replied:
I do have to say I'd prefer `.l10n.php` over `.array.php` as it feels weird to put the format into the file extensions, and I don't envision us using multiple different formats. And if we ever do, then we can just check the return value to know the type, we don't need the file extension for hat.

swissspidy replied:
A PR for `.l10n.php`: https://github.com/swissspidy/wordpress-develop/pull/50

ocean90 commented on src/wp-includes/i18n/class-wp-i18n-translations.php:
```diff
@@ -0,0 +1,148 @@
+	 */
+	public function __get( string $name ) {
+		if ( 'entries' === $name ) {
+			$entries = WP_I18n_Translation_Controller::instance()->get_entries( $this->textdomain );
```
Thoughts on passing an instance of WP_I18n_Translation_Controller via the constructor to the class so you can just use $this->controller->… (read dependency injection)

swissspidy replied:
We could do it, but isn't that a bit pointless when `WP_I18n_Translation_Controller` is a singleton?

felixarntz replied:
It makes this class more testable, so +1 to injecting via constructor. I don't think we need to use a singleton, as in it should be possible to create instances of `WP_I18n_Translation_Controller`. The `instance()` method makes sense the way things currently work, but I don't think we should make the constructor private.

swissspidy replied:
Well, here's a PR implementing your suggestion: https://github.com/swissspidy/wordpress-develop/pull/49

Is that what you had in mind?

ocean90 replied:
From only looking at the changes, yes! I know that core loves singletons, but that shouldn't prevent us from writing new code in a more modern way 🙃

swissspidy committed:
Fix `@subpackage` annotations

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Use loop, update docs

swissspidy committed:
Pass i18n controller instance to `WP_I18N_Translations`

ocean90 commented on src/wp-includes/i18n/class-wp-i18n-translation-file-mo.php:
```diff
@@ -0,0 +1,221 @@
+ *
+ * @since 6.5.0
+ */
+class WP_I18n_Translation_File_MO extends WP_I18n_Translation_File {
```
```suggestion
class WP_Translation_File_MO extends WP_Translation_File {
```

I think we can simplify the naming of all the classes by removing the `I18n` part. Not only because `I18n` reads bad (`I18N` would be my preference) but also because I can't see what other `WP_*_Translation_File` classes core would need. Thoughts?

swissspidy replied:
I think I meant `I18N` but no objections to dropping that prefix 👍

ocean90 commented:
> If an MO file has a corresponding PHP file, that file is loaded instead.

To keep the PR even "simpler" can we remove this part from this PR for now too? So basically as a first step just let add the foundation (replacement of POMO) and then extend that with the PHP format. This should give us more confidence that the replacement is working as expected with just MO which will still likely be the primary use case for a while.

swissspidy commented:
@ocean90 What would be your proposed timeline for that? For gradual adoption I would recommend not removing the part but simply changing the default format in the `translation_file_format` filter to `mo`. Later on we can then simply switch the default to `php`. Given the benefits of PHP I would suggest doing that sooner rather than later. Alternatively, keep the default but don't ship PHP translation files yet (so control it via the translation API)

swissspidy committed:
Drop `_i18n_` infix

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

felixarntz commented on src/wp-includes/l10n.php:
```diff
@@ -866,8 +908,11 @@ function unload_textdomain( $domain, $reloadable = false ) {
 
 		unset( $l10n[ $domain ] );
 
+		// Since multiple locales are supported,reloadable text domains don't actually need to be unloaded.
```
```suggestion
		// Since multiple locales are supported, reloadable text domains don't actually need to be unloaded.
```

felixarntz commented on src/wp-includes/i18n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string                          $textdomain Optional. Text domain. Default 'default'.
+	 * @param WP_Translation_File|string $mo         Translation file instance or file name.
+	 * @param string                          $locale     Optional. Locale. Default all locales.
```
```suggestion
	 * @param string                     $textdomain Optional. Text domain. Default 'default'.
	 * @param WP_Translation_File|string $mo         Translation file instance or file name.
	 * @param string                     $locale     Optional. Locale. Default all locales.
```

felixarntz commented on src/wp-includes/i18n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+				);
+			}
+			if ( false !== $moe->error() ) {
+				// Unload this file, something is wrong.
+				$this->unload( $textdomain, $moe, $locale );
```
Wouldn't we have to call `$l10n_unloaded[ $domain ] = true;` here too for proper compat with that global?

swissspidy replied:
While tese share a similar name they're not that related.

Tis controller is self-contained and does not deal with any of the WordPress globals.

Plus this controller holds translations for multiple files and domains, just because one file for a domain is not working anymore does not mean the whole domain needs to be marked as unloaded.

felixarntz replied:
@swissspidy I am asking because you set `$l10n_unloaded[ $domain ] = true;` in the other place where you call the `unload()` method? So why is this done in one place but not the other?

swissspidy replied:
That other place is within `unload_textdomain()` where this was always there and not a new addition. Here we are in the new translation controller which operates outside of that.

As I said, it is a pretty self-contained library and does not deal with WordPress globals. And even if it would, it also supports multiple translation files _per domain_. And just because loading one file of many failed, it does not mean the whole domain could be marked as unloaded.

Simply put, there is no need to put any `$l10n_unloaded[ $domain ] = true;` calls into this class here as it doesn't make sense.

And if you are wondering what this `$this->unload( $textdomain, $moe, $locale );` line does:

This controller/library is smart and doesn't actually parse translation files until the first translation is looked up. But that also means after loading the translation and looking up the translation, it could detect that the translation file is somehow corrupt, and then remove it from the stack.

That still does not warrant setting that global.

felixarntz commented on src/wp-includes/i18n/class-wp-translations.php:
```diff
@@ -0,0 +1,159 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_Translation_Controller $controller I18N controller.
+	 * @param string                         $textdomain Text domain.
```
```suggestion
	 * @param WP_Translation_Controller $controller I18N controller.
	 * @param string                    $textdomain Text domain.
```

swissspidy committed:
Docblock improvements

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Revert accidental change

swissspidy committed:
Rename test files

swissspidy committed:
Add missing parent call

swissspidy commented:
btw, yes I am aware of the failing tests. And no, leaky core tests are not fun.

swissspidy committed:
Simplify unloading

swissspidy committed:
Rename parameter

swissspidy committed:
Improve cleanup in other test

swissspidy committed:
Avoid leaky tests with multiple interfering class instances

swissspidy committed:
move unload higher up

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Add test for e572b0c234d1d5cdf838b00637e0bb945474fb73

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Rename `.mo.php` to `.l10n.php`

swissspidy committed:
Docblock improvements

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo-fix-ext

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Add text domain to `translation_file_format` filter

akirk reviewed (commented):
Sorry, moved my comment to https://core.trac.wordpress.org/ticket/59656#comment:11

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

felixarntz commented on src/wp-includes/i18n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @var array<string,array<string, array<string, WP_Translation_File|false>>>
+	 */
+	protected $loaded_files = array();
```
Nit-pick: While not mandated by WPCS, per WP coding standards there shouldn't be a blank line between `@since` and `@var` on class properties. See https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#2-classes:~:text=2.1%20Class%20Members-,2.1.1%20Properties,-Class%20properties%20should
```suggestion
	/**
	 * Current locale.
	 *
	 * @since 6.5.0
	 * @var string
	 */
	protected $current_locale = 'en_US';

	/**
	 * Map of loaded translations per locale and text domain.
	 *
	 * [ Locale => [ Textdomain => [ .., .. ] ] ]
	 *
	 * @since 6.5.0
	 * @var array<string, array<string, WP_Translation_File[]>>
	 */
	protected $loaded_translations = array();

	/**
	 * List of loaded translation files.
	 *
	 * [ Filename => [ Locale => [ Textdomain => WP_Translation_File ] ] ]
	 *
	 * @since 6.5.0
	 * @var array<string,array<string, array<string, WP_Translation_File|false>>>
	 */
	protected $loaded_files = array();
```

felixarntz commented on src/wp-includes/i18n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,404 @@
+	 * @param string                     $locale     Optional. Locale. Default all locales.
+	 * @return bool True on success, false otherwise.
+	 */
+	public function unload( string $textdomain = 'default', $file = null, string $locale = null ) {
```
I find this a bit of a clunky API for 2 reasons:
* The order of parameters is different than that in `load()`, which is confusing. I realize that's probably the case because `$file` is optional, but see my point below on that itself being an issue.
* The method supports two distinct things based on whether you pass a `$file` or not, which is an antipattern. This is further confirmed by the method logic being entirely separate just based on whether that parameter is passed.

I think we should have these instead:
* `unload( string $translation_file, string $textdomain = 'default', string $locale = null )`: Unloads a single translation file and thus is the logical counterpart to the `load()` method, with similar parameters.
* `unload_textdomain( string $textdomain = 'default', string $locale = null )`: Covers the use-case of unloading an entire textdomain. We could consider another name like `unload_all()`, but my initial hunch is `unload_textdomain()` is more clear.

Alternatively, if we wanted to make things more verbose (benefit: more clear, downside: maybe a bit long), we could use `load_file()` and `unload_file()` instead of just `load()` and `unload()`, to make it very obvious which methods are for a single translation file vs a whole textdomain.

swissspidy replied:
Went with `load_file` / `unload_file` / `unload_textdomain`

Leaves the option to add `unload_all` in the future to truly unload everything, without having to pass any arguments.

felixarntz commented on src/wp-includes/i18n/class-wp-translation-file-mo.php:
```diff
@@ -0,0 +1,221 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var int
```
See this and the other classes' properties as well regarding the additional blank line :)

felixarntz commented on src/wp-includes/i18n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,302 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var bool|string
```
Nit-pick: Why is this `bool|string`. Maybe a bit cleaner to always be a string? An empty string still signifies no errors.

swissspidy replied:
It was already like that when I took over the library. Mixing two types is indeed not ideal, though in this case I prefer `null|string` over empty vs non-empty strings

felixarntz commented on src/wp-includes/i18n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,302 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return bool|string Error
```
See above - also missing period :)

I would suggest:
```suggestion
	 * @return string Error text, or empty string if no error.
```

felixarntz commented on src/wp-includes/i18n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,302 @@
+	 * @param string $filetype Desired target file type.
+	 * @return string|false Transformed translation file contents on success, false otherwise.
+	 */
+	public static function transform( string $file, string $filetype ) {
```
Nit-pick: Maybe colocate all the static methods either at the top or bottom of the class? I think it makes it a bit more intuitive to scan the class and see which static methods are present, as their responsibility is quite different from the instance methods.

felixarntz reviewed (approved):
@swissspidy This looks great to me!

I left a few minor comments that would be great to address, but overall I think this is almost good to go.

swissspidy committed:
Merge pull request #50 from swissspidy/add/ginger-mo-fix-ext

Rename `.mo.php` to `.l10n.php`

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Minor adjustments

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy committed:
Another minor docblock change

felixarntz commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $locale     Optional. Locale. Default current locale.
+	 * @return bool True if there are any loaded translations, false otherwise.
+	 */
+	public function is_loaded( string $textdomain = 'default', string $locale = null ) {
```
Should we call this `is_textdomain_loaded()` now that we added the relevant noun to the other related methods in this class?

westonruter replied:
```suggestion
	public function is_loaded( string $textdomain = 'default', string $locale = null ): bool {
```

felixarntz reviewed (commented):
Thanks for the updates @swissspidy, looks great!

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

westonruter commented on src/wp-includes/l10n.php:
```diff
@@ -797,22 +797,65 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+	if ( 'mo' !== $preferred_format ) {
+		array_unshift(
+			$translation_files,
+			str_replace( '.mo', ".l10n.$preferred_format", $mofile )
```
Is there a chance that `.mo` will appear elsewhere in `$mofile`, e.g. not in the file extension?

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * [ Filename => [ Locale => [ Textdomain => WP_Translation_File ] ] ]
+	 *
+	 * @since 6.5.0
+	 * @var array<string,array<string, array<string, WP_Translation_File|false>>>
```
😍 and yet nit:

```suggestion
	 * @var array<string, array<string, array<string, WP_Translation_File|false>>>
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+		static $instance;
+
+		if ( ! $instance ) {
+			$instance = new WP_Translation_Controller();
```
Would this be better?

```suggestion
			$instance = new self();
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	/**
+	 * Map of loaded translations per locale and text domain.
+	 *
+	 * [ Locale => [ Textdomain => [ .., .. ] ] ]
```
what does `..` signify? Is it `...`?

```suggestion
	 * [ Locale => [ Textdomain => [ ..., ... ] ] ]
```

swissspidy replied:
It signifies a placeholder, so yes, I guess three dots works as well

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+ *
+ * @since 6.5.0
+ */
+class WP_Translation_Controller {
```
Does this need to be extended? If not:

```suggestion
final class WP_Translation_Controller {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 *
+	 * @return WP_Translation_Controller
+	 */
+	public static function instance() {
```
Since PHP 7.0 types are being used in this class (🎉), should the return types be added as well?

```suggestion
	public static function instance(): WP_Translation_Controller {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 *
+	 * @return string Locale.
+	 */
+	public function get_locale() {
```
```suggestion
	public function get_locale(): string {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $locale           Optional. Locale. Default current locale.
+	 * @return bool True on success, false otherwise.
+	 */
+	public function load_file( string $translation_file, string $textdomain = 'default', string $locale = null ) {
```
```suggestion
	public function load_file( string $translation_file, string $textdomain = 'default', string $locale = null ): bool {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string                     $locale     Optional. Locale. Default all locales.
+	 * @return bool True on success, false otherwise.
+	 */
+	public function unload_file( $file, string $textdomain = 'default', string $locale = null ) {
```
```suggestion
	public function unload_file( $file, string $textdomain = 'default', string $locale = null ): bool {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param WP_Translation_File|string $file       Optional. Translation file instance or file name. Default all files.
+	 * @param string                     $textdomain Optional. Text domain. Default 'default'.
+	 * @param string                     $locale     Optional. Locale. Default all locales.
+	 * @return bool True on success, false otherwise.
```
This actually never returns `false`. Does it need to return anything then?

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+			}
+		}
+
+		return true;
```
Was this meant to return `false`?

```suggestion
		return false;
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $locale     Optional. Locale. Default all locales.
+	 * @return bool True on success, false otherwise.
+	 */
+	public function unload_textdomain( string $textdomain = 'default', string $locale = null ) {
```
```suggestion
	public function unload_textdomain( string $textdomain = 'default', string $locale = null ): bool {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param array{0: string, 1: string} $plurals    Pair of singular and plural translations.
```
This is an array shape (which I love) but should `@type` convention be used (as well)? cf. https://wordpress.slack.com/archives/C5VCTJGH3/p1705428874962969

swissspidy replied:
Does not really work for arbitrary key-value pairs

westonruter replied:
I think the convention in core has been to do this:

```suggestion
	 * @param array {
	 *     Pair of singular and plural translations.
	 *
	 *     @type string $0 Singular translation.
	 *     @type string $1 Plural translation.
	 * }
```

For example `wp_get_attachment_image_src()`:

https://github.com/WordPress/wordpress-develop/blob/0b7165e60d4c0d8c22c1fd736ed8a5335a8f4439/src/wp-includes/media.php#L959-L966

westonruter replied:
Combining the two:


```suggestion
	 * @param array{0: string, 1: string} {
	 *     Pair of singular and plural translations.
	 *
	 *     @type string $0 Singular translation.
	 *     @type string $1 Plural translation.
	 * }
```

swissspidy replied:
~@westonruter Those are all for tuples, not key -> value pairs~

Edit: just had to re-read to realize this is a tuple here too 🤦  sorry!

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+			}
+		}
+
+		/* @var WP_Translation_File $source */
```
```suggestion
		/** @var WP_Translation_File $source */
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $textdomain Optional. Text domain. Default 'default'.
+	 * @return array<string, string> Headers.
+	 */
+	public function get_headers( string $textdomain = 'default' ) {
```
```suggestion
	public function get_headers( string $textdomain = 'default' ): array {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $header Header name.
+	 * @return string Normalized header name.
+	 */
+	protected function normalize_header( string $header ) {
```
```suggestion
	protected function normalize_header( string $header ): string {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $textdomain Optional. Text domain. Default 'default'.
+	 * @return array<string, string> Entries.
+	 */
+	public function get_entries( string $textdomain = 'default' ) {
```
```suggestion
	public function get_entries( string $textdomain = 'default' ): array {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $singular   Singular translation.
+	 * @param string $textdomain Optional. Text domain. Default 'default'.
+	 * @param string $locale     Optional. Locale. Default current locale.
+	 * @return array{source: WP_Translation_File, entries: string[]}|false Translations on success, false otherwise.
```
Another array shape (❤️). Perhaps add the `@type` convention as well? cf. https://wordpress.slack.com/archives/C5VCTJGH3/p1705428874962969

westonruter commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,410 @@
+	 * @param string $locale     Optional. Locale. Default current locale.
+	 * @return WP_Translation_File[] List of translation files.
+	 */
+	protected function get_files( string $textdomain = 'default', string $locale = null ) {
```
```suggestion
	protected function get_files( string $textdomain = 'default', string $locale = null ): array {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-mo.php:
```diff
@@ -0,0 +1,219 @@
+	 *
+	 * @return bool True on success, false otherwise.
+	 */
+	protected function parse_file() {
```
```suggestion
	protected function parse_file(): bool {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-mo.php:
```diff
@@ -0,0 +1,219 @@
+		foreach ( array_keys( $entries ) as $original ) {
+			$o_addr        .= pack( $this->uint32 . '*', strlen( $original ), $entry_offsets );
+			$entry_offsets += strlen( $original ) + 1;
+			$o_entries     .= $original . pack( 'x' );
```
PhpStorm is warning about this lacking the `$values` parameter.

![image](https://github.com/WordPress/wordpress-develop/assets/134745/78dc0b06-6f01-43a3-b024-e6d4088a1974)

Is this being used just to obtain a null byte? If so, should this instead be better as:

```suggestion
			$o_entries     .= $original . "\0";
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-mo.php:
```diff
@@ -0,0 +1,219 @@
+		foreach ( $entries as $translations ) {
+			$t_addr        .= pack( $this->uint32 . '*', strlen( $translations ), $entry_offsets );
+			$entry_offsets += strlen( $translations ) + 1;
+			$t_entries     .= $translations . pack( 'x' );
```
Per above:

```suggestion
			$t_entries     .= $translations . "\0";
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-mo.php:
```diff
@@ -0,0 +1,219 @@
+	 *
+	 * @return string Translation file contents.
+	 */
+	public function export() {
```
```suggestion
	public function export(): string {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-php.php:
```diff
@@ -0,0 +1,83 @@
+
+		$result = include $this->file;
+		if ( ! $result || ! is_array( $result ) ) {
+			$this->error = true;
```
According to the phpdoc, `\WP_Translation_File::$error` is `string|null`. So should this be changed to a string?

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-php.php:
```diff
@@ -0,0 +1,83 @@
+	 *
+	 * @return string Translation file contents.
+	 */
+	public function export() {
```
```suggestion
	public function export(): string {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file-php.php:
```diff
@@ -0,0 +1,83 @@
+	 * @param mixed $value The variable you want to export.
+	 * @return string The variable representation.
+	 */
+	private function var_export( $value ) {
```
```suggestion
	private function var_export( $value ): string {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 *
+	 * @return array<string, string> Headers.
+	 */
+	public function headers() {
```
```suggestion
	public function headers(): array {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 *
+	 * @return array<string, string[]> Entries.
+	 */
+	public function entries() {
```
```suggestion
	public function entries(): array {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return array<string, string[]> Entries.
```
Is `array<string, string[]>` right here? It is not consistent with the type for `WP_Translation_File::$entries`, which is `array<string, string>`.

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 *
+	 * @return string File name.
+	 */
+	public function get_file() {
```
```suggestion
	public function get_file(): string {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 * @param string $text String to translate.
+	 * @return false|string Translation(s) on success, false otherwise.
+	 */
+	public function translate( string $text ) {
```
```suggestion
	public function translate( string $text ): bool {
```

swissspidy replied:
Incorrect, as it can return false or string

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 * @param int $number Count.
+	 * @return int Plural form.
+	 */
+	public function get_plural_form( int $number ) {
```
```suggestion
	public function get_plural_form( int $number ): int {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 * @param string $expression Plural form expression.
+	 * @return callable(int $num): int Plural forms function.
+	 */
+	public function make_plural_form_function( string $expression ) {
```
```suggestion
	public function make_plural_form_function( string $expression ): callable {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 * @param WP_Translation_File $source Source file.
+	 * @return bool True on success, false otherwise.
+	 */
+	protected function import( WP_Translation_File $source ) {
```
```suggestion
	protected function import( WP_Translation_File $source ): bool {
```

westonruter commented on src/wp-includes/l10n/class-wp-translation-file.php:
```diff
@@ -0,0 +1,296 @@
+	 * Translation entries.
+	 *
+	 * @since 6.5.0
+	 * @var array<string, string>
```
I think this is what was intended:

```suggestion
	 * @var array<string, string[]>
```

westonruter commented on src/wp-includes/l10n.php:
```diff
@@ -797,22 +797,65 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+		 * @param string $file   Path to the translation file to load.
+		 * @param string $domain The text domain.
+		 */
+		$file = apply_filters( 'load_translation_file', $file, $domain );
```
```suggestion
		$file = (string) apply_filters( 'load_translation_file', $file, $domain );
```

westonruter reviewed (changes requested):
I fully admit I have almost no experience with MO files and the I18N component, but I think I found some general things that might be improved.

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy commented:
@westonruter For the return types I wasn't actually sure whether they are allowed in core, which is why I removed them at one point.

But seems like they are: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#type-declarations - just the docs need a little updating now that PHP 7 is the required minimum

swissspidy committed:
Add back return types

Co-authored-by: Weston Ruter <westonruter@google.com>

swissspidy committed:
Rename method

swissspidy committed:
Remove errant return type again

swissspidy committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

swissspidy committed:
Lint fix

swissspidy committed:
Return false if nothing unloaded

swissspidy committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

swissspidy committed:
use `substr_replace`

swissspidy committed:
Add return type

swissspidy committed:
fix phpdoc

swissspidy committed:
String errors

swissspidy committed:
Array shape

audrasjb committed:
Bundled Themes: Typo corrections in various themes docblocks.

Props shailu25.
Fixes #60268.
See #59651.





git-svn-id: https://develop.svn.wordpress.org/trunk@57300 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
Update error messages in tests

audrasjb committed:
Twenty Twenty: Replace `wp_date()` with `date_i18n()`.

Since WordPress 5.3 it is recommended to use `wp_date()` instead of `date_i18n()`. This changeset replaces the function in Twenty Twenty, with a fallback to `date_i18n()` for old versions of WordPress.

For more info, see https://make.wordpress.org/core/2019/09/23/date-time-improvements-wp-5-3/.

Props sachyya-sachet, joyously, sabernhardt, poena.
Fixes #48589.





git-svn-id: https://develop.svn.wordpress.org/trunk@57301 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Fix var types of parameters in `sanitize_option()` and `sanitize_option_{$option}`.

The related docblocks were previously defining `$value` and `$original_value` as if they were of type `string`, but they can also be of other types, like `array`.

Props gerardreches, crstauf, mukesh27.
Fixes #60214.
See #59651.





git-svn-id: https://develop.svn.wordpress.org/trunk@57302 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
I18N: Prevent PHP warning in `WP_Textdomain_Registry`.

Prevents a warning upon cache invalidation after language pack updates if the arguments don’t have the expected format.

Follow-up to [57287], [57290], [57298], [57299].

See #58919.

git-svn-id: https://develop.svn.wordpress.org/trunk@57303 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Twenty Twenty-Four: Remove extra tab character inside the text domain.

Follow-up to [57281].

Props sabernhardt.
Fixes #60245.

git-svn-id: https://develop.svn.wordpress.org/trunk@57304 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
I18N: Fix duplicate `determine_locale()` tests added in [57286].

Props johnbillion.
See #58696.

git-svn-id: https://develop.svn.wordpress.org/trunk@57305 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
Apply suggestion

Co-authored-by: Weston Ruter <westonruter@google.com>

swissspidy committed:
Fix phpdoc

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

SergeyBiryukov commented on src/wp-includes/compat.php:
```diff
@@ -497,3 +497,35 @@ function str_ends_with( $haystack, $needle ) {
 	define( 'IMG_WEBP', IMAGETYPE_WEBP );
 }
+
+if ( ! function_exists( 'array_is_list' ) ) {
```
Could we move this next to `array_key_first()` and `array_key_last()`, to have the array-related functions grouped together for a bit more consistent placement?

SergeyBiryukov commented on src/wp-includes/l10n.php:
```diff
@@ -797,22 +797,65 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+				$i18n_controller->load_file( $l10n[ $domain ]->get_filename(), $domain, $locale );
+			}
+
+			// Unset Noop_Translations reference in get_translations_for_domain.
```
```suggestion
			// Unset NOOP_Translations reference in get_translations_for_domain().
```
Correcting the class name capitalization and adding brackets for the function name.

SergeyBiryukov commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,420 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_Translation_File|string $file       Optional. Translation file instance or file name. Default all files.
+	 * @param string                     $textdomain Optional. Text domain. Default 'default'.
+	 * @param string                     $locale     Optional. Locale. Default all locales.
```
```suggestion
	 * @param WP_Translation_File|string $file       Optional. Translation file instance or file name. Defaults to all files.
	 * @param string                     $textdomain Optional. Text domain. Default 'default'.
	 * @param string                     $locale     Optional. Locale. Defaults to all locales.
```
Unless it's an actual default value, e.g. a specific string, I think we use the `Defaults to ...` wording more often, e.g. `Defaults to the current post`.

Additionally,  the `$file` parameter appears to be required, not optional.

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

SergeyBiryukov commented on src/wp-includes/l10n/class-wp-translation-controller.php:
```diff
@@ -0,0 +1,420 @@
+	 * @since 6.5.0
+	 *
+	 * @param string $textdomain Optional. Text domain. Default 'default'.
+	 * @param string $locale     Optional. Locale. Default all locales.
```
```suggestion
	 * @param string $locale     Optional. Locale. Defaults to all locales.
```

swissspidy committed:
Move compat function closer to others

swissspidy committed:
Correcting the class name capitalization and adding brackets for the function name

swissspidy committed:
Update docblocks as per feedback

swissspidy committed:
Fix description

swissspidy committed:
Do not change `temp_filename` signature

swissspidy committed:
Merge branch 'trunk' into add/ginger-mo

swissspidy commented:
Committed in https://core.trac.wordpress.org/changeset/57337 💥 🎉 🚀

mukeshpanchal27 commented on src/wp-includes/l10n/class-wp-translations.php:
```diff
@@ -0,0 +1,157 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $original    Original string to translate from MO file. Might contain
+	 *                            0x04 as context separator or 0x00 as singular/plural separator.
+	 * @param string $translations Translation strings from MO file.
```
nit: indention is not correct for docblock.

mukeshpanchal27 commented on tests/phpunit/tests/l10n/wpTranslationsConvert.php:
```diff
@@ -0,0 +1,598 @@
+	 *
+	 * @return void
+	 */
+	public function test_get_instance() {
```
All tests should have the `@ticket` annotation.

🔢 applies to all tests

swissspidy replied:
We typically only add these for bugfixes, not to _any_ test

mukeshpanchal27 commented on tests/phpunit/tests/l10n/wpTranslationsConvert.php:
```diff
@@ -0,0 +1,598 @@
+class WP_Translation_Controller_Convert_Tests extends WP_UnitTestCase {
+	/**
+	 * @covers ::instance
+	 *
+	 * @return void
```
Please omit any `@return void`.

Check ref: https://github.com/WordPress/performance/pull/529#discussion_r982799783

swissspidy commented:
@mukeshpanchal27 thanks for the feedback. just fyi it's easy to miss feedback on closed PRs, I almost didn't see your comments because I didn't get any notification. Best to ping separately.

mukeshpanchal27 commented:
Yes, but it's nit-pick feedback, so I missed pinging you.
